internal/controller/nonadmindownloadrequest_controller.go

```diff
@@ -0,0 +1,308 @@
+	if err := r.Delete(ctx, req); err != nil {
+		if !apierrors.IsNotFound(err) {
+			logger.V(1).Info("Failed to delete expired NonAdminDownloadRequest", req.Name, err)
+			// other errors, requeue to retry delete
+			return err
+		}
+		// not found, stop requeue
+	}
+	return nil
+}
+
+// if url is available and not expired, then set status to completed
+func (r *NonAdminDownloadRequestReconciler) updateNADRWithURL(ctx context.Context, veleroDR *velerov1.DownloadRequest, req *nacv1alpha1.NonAdminDownloadRequest) error {
+	logger := log.FromContext(ctx)
+	if veleroDR.Status.DownloadURL != constant.EmptyString &&
+		veleroDR.Status.Phase == velerov1.DownloadRequestPhaseProcessed &&
+		!veleroDR.Status.Expiration.Before(&metav1.Time{Time: time.Now()}) {
+		// copy req, prepare to patch
+		prePatch := req.DeepCopy()
+		// copy status to NADR from VDR
+		req.Status.VeleroDownloadRequest.Status = &veleroDR.Status
+		// veleroDR is processed, update NADR status
+		if patchErr := r.Status().Patch(ctx, req, client.MergeFrom(prePatch)); patchErr != nil {
+			logger.Error(patchErr, "unable to patch status", req.Kind, req.Name)
+			return patchErr
+		}
+		if patchErr := r.patchAddStatusTrueConditionPhase(ctx, req, nacv1alpha1.NonAdminPhaseCreated, nacv1alpha1.ConditionNonAdminProcessed, constant.EmptyString, "Success"); patchErr != nil {
+			logger.Error(patchErr, "unable to patch status", req.Kind, req.Name)
+			return patchErr
+		}
+	}
+	return nil
+}
+
+// processDownloadRequest creates download requests after performing NaBSL checks on NAB/NAR from the downloadRequest
+func (r *NonAdminDownloadRequestReconciler) processDownloadRequest(ctx context.Context, veleroDR *velerov1.DownloadRequest, req *nacv1alpha1.NonAdminDownloadRequest) error {
+	logger := log.FromContext(ctx)
+	var nab nacv1alpha1.NonAdminBackup // holds nonadminbackup
+	var nabName string                 // holds nonadminbackup name
+	switch req.Spec.Target.Kind {
+	case velerov1.DownloadTargetKindBackupLog,
+		velerov1.DownloadTargetKindBackupContents,
+		velerov1.DownloadTargetKindBackupVolumeSnapshots,
+		velerov1.DownloadTargetKindBackupItemOperations,
+		velerov1.DownloadTargetKindBackupResourceList,
+		velerov1.DownloadTargetKindBackupResults,
+		velerov1.DownloadTargetKindCSIBackupVolumeSnapshots,
+		velerov1.DownloadTargetKindCSIBackupVolumeSnapshotContents,
+		velerov1.DownloadTargetKindBackupVolumeInfos:
+		nabName = req.Spec.Target.Name
+	case velerov1.DownloadTargetKindRestoreLog,
+		velerov1.DownloadTargetKindRestoreResults,
+		velerov1.DownloadTargetKindRestoreResourceList,
+		velerov1.DownloadTargetKindRestoreItemOperations,
+		velerov1.DownloadTargetKindRestoreVolumeInfo:
+		var nar nacv1alpha1.NonAdminRestore
+		if err := r.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: req.Spec.Target.Name}, &nar); err != nil {
+			patchErr := r.patchAddErrorStatusTrueConditionBackoff(ctx, req, nacv1alpha1.ConditionNonAdminRestoreNotAvailable, "cannot get nonadminrestore")
```

Comment:
is this right? `patchErr` is not checked if it is nil

Reply:
It is being logged.